src/java/org/apache/cassandra/db/streaming/CassandraStreamManager.java

```diff
@@ -0,0 +1,158 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.db.lifecycle.SSTableIntervalTree;
+import org.apache.cassandra.db.lifecycle.SSTableSet;
+import org.apache.cassandra.db.lifecycle.View;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.service.ActiveRepairService;
+import org.apache.cassandra.streaming.IncomingStream;
+import org.apache.cassandra.streaming.OutgoingStream;
+import org.apache.cassandra.streaming.PreviewKind;
+import org.apache.cassandra.streaming.StreamReceiver;
+import org.apache.cassandra.streaming.StreamSession;
+import org.apache.cassandra.streaming.TableStreamManager;
+import org.apache.cassandra.streaming.messages.StreamMessageHeader;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.concurrent.Ref;
+import org.apache.cassandra.utils.concurrent.Refs;
+
+/**
+ * Implements the streaming interface for the native cassandra storage engine.
+ *
+ * Handles the streaming a one or more section of one of more sstables to and from a specific
+ * remote node. The sending side performs a block-level transfer of the source stream, while the receiver
+ * must deserilaize that data stream into an partitions and rows, and then write that out as an sstable.
+ */
+public class CassandraStreamManager implements TableStreamManager
+{
+    private static final Logger logger = LoggerFactory.getLogger(CassandraStreamManager.class);
+
+    private final ColumnFamilyStore cfs;
+
+    public CassandraStreamManager(ColumnFamilyStore cfs)
+    {
+        this.cfs = cfs;
+    }
+
+    @Override
+    public IncomingStream prepareIncomingStream(StreamSession session, StreamMessageHeader header)
+    {
+        return new CassandraIncomingFile(cfs, session, header);
+    }
+
+    @Override
+    public StreamReceiver createStreamReceiver(StreamSession session, int totalStreams)
+    {
+        return new CassandraStreamReceiver(cfs, session, totalStreams);
+    }
+
+    private static Predicate<SSTableReader> getPreviewPredicate(PreviewKind kind)
+    {
+        switch (kind)
+        {
+            case ALL:
+                return Predicates.alwaysTrue();
+            case UNREPAIRED:
+                return Predicates.not(SSTableReader::isRepaired);
+            case REPAIRED:
+                return SSTableReader::isRepaired;
+            default:
+                throw new IllegalArgumentException("Unsupported kind: " + kind);
+        }
+    }
+
+    @Override
+    public Collection<OutgoingStream> createOutgoingStreams(StreamSession session, Collection<Range<Token>> ranges, UUID pendingRepair, PreviewKind previewKind)
+    {
+        Refs<SSTableReader> refs = new Refs<>();
+        final List<Range<PartitionPosition>> keyRanges = new ArrayList<>(ranges.size());
```

Comment:
pushed up a new commit to fix. The idea was that the refs release had been moved to the caller, `StreamSession.getOutgoingStreamsForRanges`, since `StreamSession.getSSTableSectionsForRanges` worked on all tables being streamed, and the stream handler only works on one table. Still, that didn't handle the refs taken in the method, but not yet returned as streams, so we need both.